### PR TITLE
Fixed possible NullPointerException

### LINKED ISSUE #378: Calling FlexibleAdapter.clear() may cause NullPointerException
Library Version: 5.0.0-rc1

All methods referenced are located in FlexibleAdapter.java

Expected outcome:
When `public IHeader getHeaderOf(@NonNull T item) { ... }` is overriden, `item` should never be null. I relied on that null checking, and it caused a NullPointerException unexpectedly.

Actual outcome:
`clear();`
calls
`removeRange(0, getItemCount(), null);`
which then assigns
`T item = getItem(positionStart);`
without checking if the `itemCount` argument is `0`.
Since the list is empty, `getItem(0);` returns null, which is then passed to `getHeaderOf(item);` without a null check.

### PR DESCRIPTION
**Pull request by magneticflux-**
Added null checks, extra annotations, and a test case.

Closes #378

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -14,7 +14,6 @@ gen/
 
 # Gradle files
 .gradle/
-gradle/
```
Comment (magneticflux-) [inline .gitignore]: @davideas I believe that this line should be removed to allow the wrapper to be included, but I just want to check that this won't break anything. Gradle says to include the wrapper files in repositories [here](https://docs.gradle.org/current/dsl/org.gradle.api.tasks.wrapper.Wrapper.html), and there are already versions of the file in the repository; the problem is, they aren't updated by a simple `git push` after changing one of them.

Comment (davideas) [inline .gitignore]: The plugin I'm using, put that.
But the wrapper doesn't depends by the local library? isn't auto-generated then?

**Comment by davideas** [comment]
@magneticflux- If there aren't any changes more, I will merge this by Friday evening.
Thanks for reviewing it.

File: .gitignore (hunk 1 above)
Comment (magneticflux-) [inline .gitignore]: The wrapper is *not* auto-generated, users must manually download the wrapper if they wish to build the project. If the version specified in the `gradle-wrapper.properties` file is changed, the current wrapper will update itself to the new version. Well, regardless of if this line is present in the `.gitignore`, the files are present in the repository. All this line is doing currently is preventing any updates to the Gradle Wrapper version. This change shouldn't affect anything right now, only allow for future updates of the Gradle version.

**Pull request merged**